src/main/scala/scalismo/mesh/PolyLine.scala

```diff
@@ -0,0 +1,171 @@
+    val A = pointSet.point(t.ptId1)
+    val B = pointSet.point(t.ptId2)
+    (B - A).norm
+  }
+}
+
+object PolyLine {
+
+  /** Typeclass for creating domains of arbitrary dimensionality */
+  trait Create[D <: Dim] {
+    def createPolyLine(pointSet: UnstructuredPointsDomain[D], topology: LineList): PolyLine[D]
+  }
+
+  implicit object Create2D extends Create[_2D] {
+    override def createPolyLine(pointSet: UnstructuredPointsDomain[_2D], topology: LineList) = {
+      PolyLine2D(pointSet, topology)
+    }
+  }
+
+  implicit object Create3D extends Create[_3D] {
+    override def createPolyLine(pointSet: UnstructuredPointsDomain[_3D], topology: LineList) = {
+      PolyLine3D(pointSet, topology)
+    }
+  }
+
+  implicit def parametricToConcreteType2D(polyLine: PolyLine[_2D]): PolyLine2D = {
+    polyLine.asInstanceOf[PolyLine2D]
+  }
+
+  implicit def parametricToConcreteType3D(polyLine: PolyLine[_3D]): PolyLine3D = {
+    polyLine.asInstanceOf[PolyLine3D]
+  }
+
+  def apply[D <: Dim](points: UnstructuredPointsDomain[D], topology: LineList)(implicit creator: Create[D]): PolyLine[D] = {
+    creator.createPolyLine(points, topology)
+  }
+
+  def enforceConsistentCellDirections[D <: Dim](polyLine: PolyLine[D])(implicit creator: PolyLine.Create[D]): PolyLine[D] = {
+
+    def reorientRecursive(curLine: LineCell, reorientedLines: IndexedSeq[LineCell]): IndexedSeq[LineCell] = {
+      if (reorientedLines.contains(curLine)) {
+        reorientedLines
+      } else {
+        val adjLines = polyLine.topology.adjacentLinesForPoint(curLine.ptId2).map(id => polyLine.topology.lines(id.id))
+
+        require(adjLines.size <= 2)
+
+        val adjLine = if (adjLines.head == curLine || adjLines.head == LineCell(curLine.ptId2, curLine.ptId1)) adjLines.last else adjLines.head
+
+        val newAdjLine = if (adjLine.ptId1 == curLine.ptId2) {
+          adjLine
+        } else {
+          LineCell(adjLine.ptId2, adjLine.ptId1)
+        }
+
+        reorientRecursive(newAdjLine, reorientedLines :+ curLine)
+      }
+    }
+    val consistentLineList = LineList(reorientRecursive(polyLine.lines.head, IndexedSeq[LineCell]()))
+    creator.createPolyLine(polyLine.pointSet, consistentLineList)
+  }
+}
+
+case class PolyLine2D(override val pointSet: UnstructuredPointsDomain[_2D], override val topology: LineList) extends PolyLine[_2D](pointSet, topology) {
+
+  /** Get all cell normals as a surface property */
+  lazy val cellNormals: LineProperty[Vector2D] = {
+    LineProperty(topology, lines.map(computeCellNormal))
+  }
+
+  /** Get all vertex normals as a surface property */
+
+  lazy val vertexNormals: ContourPointProperty[Vector2D] = {
+
+    val pd = pointSet.pointIds.map { id =>
+      val cellIds = topology.adjacentLinesForPoint(id)
+      val n = cellIds.foldLeft(Vector.zeros[_2D])((acc, cellId) => acc + computeCellNormal(topology.lines(cellId.id)))
+      n / cellIds.size: Vector2D
+    }
+
+    ContourPointProperty(topology, pd.toIndexedSeq)
+  }
+
```

Comment:
code duplication? available as `ContourPointProperty.averagedPointProperty` - although less performant, but easier to read